2016/notes-on-testing-with-jest.md

```diff
@@ -0,0 +1,280 @@
+
+```javascript
+import { runasynctest, runsynctest } from './actionsUtil';
+// fetchVal is your typical fetch action which returns a promise
+import { fetchVal } from '../my-actions';
+import * as types from 'constants/my-constants';
+
+describe('async-actions', () => {
+    runasynctest(
+    {
+        testName: 'should simulate fetch action',
+        expectedActions: [
+        { type: types.GET__REQUEST },
+        {
+            type: types.GET__SUCCESS,
+            pids: [1, 2]
+        }
+        ],
+        fn: fetchVal,
+        params: [[1, 2]]
+    });
+});
+```
+
+Components with Actions
+-----------------------
+
+If in the above example for `MyComponent`'s testing, the `handleClick` were to send actions with redux like
+
+__my-component.js__
+
+```javascript
+...
+
+handleClick = event => {
+    this.props.fetchValAction(event);
+}
+
+...
+```
+
+__my-component.test.js__
+
+```javascript
+describe('<MyComponent />', () => {
+    it('should call fetchVal action', () => {
+    const mockFetchVal = jest.mock();
+    const component = mount(<MyComponent fetchValAction={mockFetchVal}/>);
+    component.childAt(2).simulate('click');
+    expect(mockFetchVal).toHaveBeenCalled();
+    })
+})
+```
+
+tldr; Various solutions can be adopted. Choose whichever suits your environment.
+
+__How do I test asynchronous functions?__
+
+The Jest docs has a [tutorial on asynchronous testing](https://facebook.github.io/jest/docs/tutorial-async.html#content). Converting them to `Promises` works too.
+
+__How to simulate window events?__
+
+We used [jsdom](https://github.com/tmpvar/jsdom) to get a DOM that works with Node.js.
+We set it up for the project at Jest's [`setupFile`](https://facebook.github.io/jest/docs/configuration.html#setupfiles-array). 
+
+__setupFile.js__
+
+```javascript
+const jsdom = require('jsdom');
+
+const documentHTML = '<!doctype html><html><body><div id="root"></div></body></html>';
+global.document = jsdom.jsdom(documentHTML);
+global.window = document.parentWindow;
+global.window.resizeTo = (width, height) => {
+	global.window.innerWidth = width || global.window.innerWidth;
+	global.window.innerHeight = width || global.window.innerHeight;
+	global.window.dispatchEvent(new Event('resize'));
+};
+global.Promise = require.requireActual('promise');
+
+``` 
+
+Now we can simulate resize by the mock method above. So we have tests like below.
+
+__example__
+
+```javascript
+/*
+  The test below checks if snapshots were the same before and after resizes.
+  You can also test for other window events like scroll in this manner.
+*/
+it('should resize properly', () => {
+  const props = { highlights: true };
+  const component = mount(<MyComponent {...props} />, {attachTo: document.getElementById('root')});
+  global.window.resizeTo(1025);
+  expect(mountToJson(component)).toMatchSnapshot();
+  global.window.resizeTo(1020);
+  expect(mountToJson(component)).toMatchSnapshot();
+});
+```
+
+__I can't get PostCSS to work!__
```

Comment:
This sucks :( If PostCSS had a sync API, this would just work. The problem here is that Jest transforms files just-in-time when they are required; that's why they need to be synchronous.

One way this could work is to have a "preprocess-step" before you run your tests, like build all the CSS files into "css-out/" and then use moduleNameMapper to map from `my.css` to `css-out/my.css`.